### PR TITLE
feat(api): add exam utils and GET /exam/:id endpoint to new api

### LINKED ISSUE #51133: GET /exam/:id

### PR DESCRIPTION
**Pull request by moT01**
This adds the `GET /exam/:id` endpoint and the exam utils

`utils/exam.ts` is the main helper file included - it was ported over from the current API and translated to use typescript - It has a `generateRandomExam` function which creates a random exam from the exam in the database, used in `GET /exam/:id` - it also adds the `createExamResults` function, which is used in the submit exam endpoint (not included in this PR)

also adds the `exam` stuff to the prisma schema so we can work with the Exam collection

and lots of tests

Unrelated: Slightly modifies the jest config cause my computer couldn't handle the number of workers

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #51133

### CODE REVIEW

File: api/jest.config.ts [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
+
 const config: Config = {
   verbose: true,
   testRegex: '\\.test\\.ts$',
   transform: {
     '^.+\\.ts$': 'ts-jest'
-  }
+  },
+  ...(process.env.FREECODECAMP_NODE_ENV == 'development'
+    ? { maxWorkers: 5 }
```
Comment (raisedadead) [inline api/jest.config.ts:16]:
```suggestion
    ? { maxWorkers: NODE_MAX_ENV ||  whateverisdefault }
```

Could we do something like this?

File: api/src/routes/challenge.ts [hunk 2]
```diff
@@ -424,5 +429,95 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+          await fastify.prisma.user.findUniqueOrThrow({
+            where: { id: req.session.user.id },
+            select: { completedChallenges: true }
+          });
+
+        const examFromDb = await fastify.prisma.exam.findUnique({
+          where: { id }
+        });
+
+        console.log(examFromDb);
```
Comment (raisedadead) [inline api/src/routes/challenge.ts:459]:
```suggestion

```

File: api/src/routes/challenge.ts [hunk 3]
```diff
@@ -424,5 +429,95 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+        }
+
+        const randomizedExam = generateRandomExam(examFromDb);
+        const validGeneratedExamSchema = validateGeneratedExamSchema(
+          randomizedExam,
+          numberOfQuestionsInExam
+        );
+
+        console.log('validGeneratedExamSchema');
+        console.log(validGeneratedExamSchema);
```
Comment (raisedadead) [inline api/src/routes/challenge.ts:501]:
```suggestion

```

**Review by raisedadead — CHANGES_REQUESTED** [review]
Nitpicks

File: api/src/routes/challenge.ts [hunk 4]
```diff
@@ -424,5 +429,95 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
...
+            challenge => challenge.id === prerequisites[i]?.id
+          );
+
+          if (!prerequisiteCompleted) {
+            void reply.code(403);
+            return {
+              error: `You have not completed the required challenges to start the '${title}'.`
+            };
+          }
+        }
```
Comment (ShaunSHamilton) [inline api/src/routes/challenge.ts:492]: We want to confirm all _prerequisite_ (subset) challenges are in `completedChallenges` (set).

Small improvement would be to loop over `completedChallenges` - once - and confirm every `prerequisite` is included. I probably would have done something like (untested):

```js
const prerequisiteIds = prerequisites.map(p => p.id);
const complete = completedChallenges.filter(c => {
    return prerequisiteIds.includes(c.id);
});

if (complete.length !== prerequisites.length) {
  return "Error"
}
```

It does not seem any better, thinking about it more 🤷‍♂️

File: api/jest.config.ts (hunk 1 above)
Comment (moT01) [reply api/jest.config.ts]: I decided to just remove this from the PR - I will make another specific for it.

File: api/src/routes/challenge.ts (hunk 2 above)
Comment (moT01) [reply api/src/routes/challenge.ts]: fixed in latest commit

File: api/src/routes/challenge.ts (hunk 3 above)
Comment (moT01) [reply api/src/routes/challenge.ts]: fixed in latest commit

**Review by ShaunSHamilton — APPROVED** [review]
TMTLG 👍

**Review by raisedadead — APPROVED** [review]

**Pull request merged**